infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java

```diff
@@ -247,6 +261,56 @@ private void execute(
     }
   }
 
+  /**
+   * Watches failure events for the specified Pod, and deletes the Pod if a failure event is
+   * detected
+   *
+   * @param deployments the KubernetesDeployments object
+   * @param podName the name of the Pod to check events for
+   * @throws InfrastructureException
+   */
+  private void watchFailureEventAndTerminatePod(KubernetesDeployments deployments, String podName)
+      throws InfrastructureException {
+    deployments.watchEvents(
+        event -> {
+          if (podName.equals(event.getPodName()) && isPodFailureEvent(event)) {
+            try {
+              LOG.debug(
+                  "Deleting pod: '{}' due to failure event: '{}'", podName, event.getMessage());
+              deployments.delete(event.getPodName());
```

Comment:
Are you sure it does not conflict with
```java
        final Pod finished = deployments.wait(podName, WAIT_POD_TIMEOUT_MIN, POD_PREDICATE::apply);
```
?

I mean: please make sure wait is interrupted after you remove the pod.

Reply:
Sounds good. To interrupt the wait, I thought maybe `deployments.wait()` could return a `CompletableFuture<Pod>`, and call `cancel()` to interrupt the wait if needed, but I'm having trouble figuring out how to call `watch.close()` after canceling the `CompletableFuture`. I suppose `deployments.wait()` could return a Java object containing both the `CompletableFuture` and the `Watch` object, but I'm not sure if that's the way to go.
https://github.com/eclipse-che/che-server/blob/90968896473bf0172d9dc143906a91b9f89c96aa/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java#L236

Maybe I should instead create a method for `KubernetesDeployments` called something like `watchEventsAndWaitPod`, that would both watch for failure events and return if any, and, wait/watch for pod completion?